src/injected/analyzers/filter-results.ts

```diff
@@ -10,7 +10,8 @@ export const filterNeedsReviewResults = (results: ScanResults): ScanResults => {
         v =>
             v.id !== 'aria-input-field-name' &&
             v.id !== 'color-contrast' &&
-            v.id !== 'th-has-data-cells',
+            v.id !== 'th-has-data-cells' &&
+            v.id !== 'label-content-name-mismatch',
```

Comment:
It looks like the list of rules for needs review here is the same as the list provided in needsReviewRuleAnalyzerConfiguration.rules in src/ad-hoc-visualizations/needs-review/visualization.tsx. If these two sets of data are always intended to be the same, could we have them in one place to eliminate the possibility of either getting out of sync?

Reply:
It's not intended to be the same, though I don't think that fact is very clear from the naming of the function here.

The purpose of this function (`filterNeedsReviewResults`) is *not* to take a full set of axe-core results and to either filter out *or* to filter-keeping just needs-review results. Its purpose is to take the set of axe-core results *produced by the needs review visualizer's analyzer configuration*, which enables just those rules that can produce needs review results, and to *filter out results that are already expressed as automated checks failures*.

For example, the `color-contrast` rule needs to be enabled as part of the "needs review" analyzer configuration because we want any "incomplete" results it produces to be present in the "needs review" UI. However, we *don't* want the "violations" it produces to be present in the "needs review" UI, even though we need the rule enabled, since color-contrast violations are full automated checks failures. That's why `filterNeedsReviewResults` filters out color contrast violations.

The `label-content-name-mismatch` rule doesn't work the same way; it is an experimental rule that produces violations, not incompletes, when it detects problems. That means:

* We *don't* want it to run at all in the automated checks ruleset; there is nothing the rule can report that we would ever want to show up in an automated check. Therefore, it shouldn't be marked as included in "get-rules".
* We *do* want to see any violations that the rule reports as part of needs review. Therefore, we shouldn't filter out its violations in `filterNeedsReviewResults`.
* We *need PM input* on whether we want users to see anything that the rule reports as "incomplete" as part of *needs review* output.
    * Based on a quick review of [the rule's implementation](https://github.com/dequelabs/axe-core/blob/develop/lib/checks/label/label-content-name-mismatch-evaluate.js), it looks like that will occur if-and-only-if either the visible text or the aria-label of the element being tested is determined to be "not human interpretable", which axe-core's purposes here essentially means "is something like an emoji, punctuation mark, or a single character that might be being used similarly, like "x").
    * I would recommend that we probably *do* want such cases to be marked as needs review. If that's the case, `filterNeedsReviewResults` should *not* filter out incomplete results for this id.
    * If PM decides that we *do not* want such cases to be marked as needs review (ie, we want to ignore them silently), then `filterNeedsReviewResults` should filter out incomplete results for this rule similar to what it already does for `link-in-text-block`.

---

I don't think this function's naming makes it very clear what the intended behavior is; I think it would be a very good idea to improve the name and/or add a comment on the function that explains what the intended purpose of the function is.

An even better option (which would also addressing Rob's point about duplication) might be to separate out the notions of "what behavior do we want configured for each needs review rule" vs "how do we apply that configuration to filter axe result objects". I could imagine there being some sort of `NeedsReviewRuleConfiguration` that we initialize for each "needs review" rule that includes properties for "should violations be treated as need-review? should incompletes be treated as needs-review?", for this filtering logic to use those properties as input, and then for the visualization configuration to use the same configuration objects to identify which needs review rules exist and reuse that for its "analyzer rules" configuration.

I think that would be a much bigger change though; if you'd like to pursue that refactoring, I would probably lean towards doing it in a separate PR.